---
title: "236523 - Introduction to Bioinformatics - Spring 21 - HW2"
author: "Alon Hacohen & Roy Bernea"
date: "5/27/2021"
output:
  word_document: default
  pdf_document: default
editor_options:
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Roy Bernea - 316315829 ; Alon Hacohen - 311587653

# Qusetion 1

Following is the code and answers for the questions under the "Question 1" Section - as instructed in the assignment:

```{r,echo=FALSE} 
invisible(library("EnhancedVolcano"))
invisible(library("DESeq2"))
invisible(library("pheatmap"))
invisible(library("RColorBrewer"))

```


### Q1.2
The dataset represents the difference in gene expression between three different kinds of of human melanoma cells and a regular melanocytes cell (normal skin cell).

### Q1.3
```{r Q1.3}
c1<-c("GSM2344965", "GSM2344966", "GSM2344967", "GSM2344968", "GSM2344969", "GSM2344970", "GSM2344971", "GSM2344972", "GSM2344973", "GSM2344974", "GSM2344975", "GSM2344976")

c2<-c("FM","FM","FM","SK_MEL_28","SK_MEL_28","SK_MEL_28","SK_MEL_147","SK_MEL_147","SK_MEL_147","UACC_62","UACC_62","UACC_62")

c3<-c("normal","normal","normal","cancer","cancer","cancer","cancer","cancer","cancer","cancer","cancer","cancer")

coldata <-cbind(ID=c1,line_type=c2,condition=c3)

coldata
```

### Q1.4
```{r Q1.4}
c1<-c("GSM2344965","GSM2344966", "GSM2344968","GSM2344969", "GSM2344971","GSM2344972", "GSM2344974","GSM2344975")

c2<-c("FM","FM","SK_MEL_28","SK_MEL_28","SK_MEL_147","SK_MEL_147","UACC_62","UACC_62")

c3<-c("normal","normal","cancer","cancer","cancer","cancer","cancer","cancer")

coldata <-cbind(ID=c1,line_type=c2,condition=c3)

cts_prime <- as.matrix(read.csv("GSE88741-GPL16791_series_matrix.txt", skip = 72, header=T,sep="\t",row.names="ID_REF"))

cts<-cts_prime[,c("GSM2344965","GSM2344966", "GSM2344968","GSM2344969", "GSM2344971","GSM2344972", "GSM2344974","GSM2344975")]

cts<-cts[complete.cases(cts), ] 
dds <- DESeqDataSetFromMatrix(countData = cts, 
                              colData = coldata, 
                              design = ~ condition) 
dds <- dds[rowSums(counts(dds)) > 1,] 
dds_analysis <- DESeq(dds) 
res <- results(dds_analysis)

res
```

### Q1.5
```{r Q1.5}
EnhancedVolcano(res,lab = rownames(res),x='log2FoldChange',y='padj',FCcutoff = 0.1,xlim=c(-5,5), ylim=c(0,20), title = 'diff analysis') 
```

### Q1.6
```{r Q1.6}
dds_vst <- vst(dds, blind = FALSE)
top_genes <- head(order(rowVars(assay(dds_vst)), decreasing=TRUE), 30)
top_genes_matrix<- assay(dds_vst)[top_genes,]
row_means_mat<-rowMeans(top_genes_matrix)
top_genes_matrix<-top_genes_matrix-row_means_mat
top_genes_matrix<-t(top_genes_matrix)
top_genes_matrix<-scale(top_genes_matrix)
top_genes_matrix<-t(top_genes_matrix)
col_row_names <-as.data.frame(colData(dds_vst)[, c("line_type","condition")])
pheatmap(top_genes_matrix, annotation_col=col_row_names, x= 'padj<0.05', y='lfcThreshold=3')
```

### Q1.7
```{r Q1.7}
res_sub1 <- subset(res,padj<0.05)
ordered<-head(res_sub1[order(res_sub1$log2FoldChange,decreasing = TRUE),])
ordered
```

### The genes we chose: 

GAGE12J (upregulated)- This gene belongs to a family expressed in a large variety of tumors whereas in normal tissues, expression is restricted to germ cells. These genes organized in clustered repeats, have a high degree of predicted sequence identity, but differ by scattered single nucleotide substitution. Their sequences contain either the antigenic peptide YYWPRPRRY or YRPRPRRY which is recognized by cytotoxic T-cells.

CYSLTR2 (downregulated)- The cysteinyl leukotrienes are important mediators of human bronchial asthma. Studies have determined that cysteinyl leukotrienes activate at least 2 receptors. This encoded receptor is a member of the superfamily of G protein-coupled receptors. It seems to play a major role in endocrine and cardiovascular systems. 


# Question 2

Following is the code and answers for the questions under the "Question 2" Section - as instructed in the assignment:

### Q2.1
Why the mapped reads are mapped preferentially to exons?
# TODO - write better
The mapped reads are mapped preferentially to exons since during the process of splicing the introns are being removed from the RNA.
In the final product (the RNA molecule) there are only exons, therefore the read are mapped to the exons.

### Q2.2
Sometimes if we map RNA-seq reads of certain tissue samples and we look on certain genes we will see that reads are also mapped to the intronic region. What can be biological cause/reason for such an observation? 
# TODO
A biological reason for reads being mapped to the intronic region during the RNA-seq process is that
#EDEN's
that when the method is polyA-selection and we also have non-coding RNA in the mRNA then we can see reads that are mapped to the intronic region.


### Q2.3
### Q2.3.1
Additional names (aliases) for the gene *MARCH7* Are:
Membrane Associated Ring-CH-Type Finger
Axotrophin
MARCH-VII
RNF177
Membrane-Associated Ring Finger (C3HC4) 7, E3 Ubiquitin Protein Ligase
RING-Type E3 Ubiquitin Transferase MARCHF7
Membrane-Associated RING Finger Protein 7
Membrane-Associated RING-CH Protein VII
E3 Ubiquitin-Protein Ligase MARCHF7
RING Finger Protein 177
MARCH7
AXOT
Membrane-Associated Ring Finger (C3HC4)
RING-Type E3 Ubiquitin Transferase MARCH7
E3 Ubiquitin-Protein Ligase MARCH7
Membrane Associated Ring Finger 7
EC 2.3.2.27
EC 6.3.2 
MARCHF7
AXO

These gene names were extracted from the following databases:
HGNC - the HUGO Gene Nomenclature Committee database
Ensembl genome database
UniProt database
NCBI's Entrez molecular sequence database
OMIM - Online Mendelian Inheritance in Man database

### Q2.3.2
If we paste the name of this gene manually into an Excel sheet, meaning we paste "MARCH7" into a cell inside the sheet, a possible problem we might encounter is that Excel would recognize the name as a date - March 7 - and refer to it as one.
In such case, instead of referring to the gene's name "MARCH7" as its ID (or as a unique identifier), Excel would mistakenly refer to it as a date - which might cause us problem later on in our data analysis process (e.g. if we want to compare the gene in our excel sheet to its appearance in a database - but R would recognize the identifiers as two different ones because of their type etc.). 




# Question 3

Following is the code and answers for the questions under the "Question 3" Section - as instructed in the assignment:

### Q3.1
```{r,echo=TRUE,eval=TRUE, warning=FALSE, message=FALSE}

library(ggplot2)
library(RColorBrewer)
library(factoextra)
library(tidyverse)
library(Rtsne)
library(dbscan)

# data preperation
#change to relative path?
sdy <- readRDS("C:/CS_Technion/236523-Intro_to_Bioinformatics/Spr21/Intro_to_BioInf-Spr21-HW2/sdy420.rds")
#print(t(sdy$expr))
data <- na.omit(t(sdy$expr))
data.scaled <- scale(data)
#view(data.scaled)
#dist.mat <- dist(data)

# preparing tsne for later
tsne = Rtsne(data.scaled)

# Clustering methods

set.seed(555)

# kmeans
km.result <- kmeans(data.scaled, centers = 3, nstart = 25)
# hierarchical clustering
hc.result <- hclust(dist(data.scaled), "ward.D")
# DBSCAN
dbscan.result <- dbscan(tsne$Y, eps = 1, minPts = 4)

#dbscan.result




#Q3.2+3.3

# Plotting the clusters

# kmeans
#k=2
km.result <- kmeans(data.scaled, centers = 2, nstart = 25)
plot(tsne$Y, col=km.result$cluster, main = "k-means - k=2")
#k=3
km.result <- kmeans(data.scaled, centers = 3, nstart = 25)
plot(tsne$Y, col=km.result$cluster, main = "k-means - k=3")
#k=4
km.result <- kmeans(data.scaled, centers = 4, nstart = 25)
plot(tsne$Y, col=km.result$cluster, main = "k-means - k=4")
#k=5
km.result <- kmeans(data.scaled, centers = 5, nstart = 25)
plot(tsne$Y, col=km.result$cluster, main = "k-means - k=5")


# hierarchical clustering
#method="ward.D"
hc.result <- hclust(dist(data.scaled), "ward.D")
plot(tsne$Y, col=hc.result$order, main = "hierarchical - method=ward.D")
plot(hc.result, cex = 0.6, hang = -1)
#method="ave"
hc.result <- hclust(dist(data.scaled), "ave")
plot(tsne$Y, col=hc.result$order, main = "hierarchical - method=average")
plot(hc.result)
#method="ward.D2"
hc.result <- hclust(dist(data.scaled), "ward.D2")
plot(tsne$Y, col=hc.result$order, main = "hierarchical - method=ward.D2")
plot(hc.result)
#method="centroid"
hc.result <- hclust(dist(data.scaled), "centroid")
plot(tsne$Y, col=hc.result$order, main = "hierarchical - method=centroid")
plot(hc.result)


# DBSCAN
#eps=1,minpts=4
dbscan.result <- dbscan(tsne$Y, eps = 1, minPts = 4)
plot(tsne$Y, col=dbscan.result$cluster, main = "DBSCAN - eps=1, minpts=4")
#eps=0.5,minpts=4
dbscan.result <- dbscan(tsne$Y, eps = 0.5, minPts = 4)
plot(tsne$Y, col=dbscan.result$cluster, main = "DBSCAN - eps=0.5, minpts=4")
#eps=1,minpts=6
dbscan.result <- dbscan(tsne$Y, eps = 1, minPts = 6)
plot(tsne$Y, col=dbscan.result$cluster, main = "DBSCAN - eps=1, minpts=6")
#eps=0.5,minpts=6
dbscan.result <- dbscan(tsne$Y, eps = 0.5, minPts = 6)
plot(tsne$Y, col=dbscan.result$cluster, main = "DBSCAN - eps=0.5, minpts=6")





plot(cutree(hc.result, k=7))

# plotting to hclust dendrogram
#plot(hc.result)


```


# Q3.3
The method in hclust() that seems to give a clustering that is more accurate to the tSNE plot is ????.

# Q3.4

# TODO - write the code for rand.index

Did you get the same or different clusters using the three techniques? Find ways to present the similarities and differences between the clusterings. If the clusters are different what can explain the difference?

Answer: the clusters are different.
reasons-
the method that each clustering technique uses is different way to cluster
the similarity is because the data is small


# Q3.5
find who belongs to each cluster
show it on a graph

```{r,echo=TRUE,eval=TRUE, warning=FALSE, message=FALSE}

sample_info <- readCSV("C:/CS_Technion/236523-Intro_to_Bioinformatics/Spr21/Intro_to_BioInf-Spr21-HW2/sample_info.csv")

# new kmeans?

# maybe with boxplot? and not ggplot?
#AGE
ggplot(sample_info, aes(x=km.result$cluster, y=AGE, group = km.result$cluster)) + geom_boxplot()
#RACE
table(sample_info$RACE, km.result$cluster)
#GENDER
table(sample_info$GENDER, km.result$cluster)

```


# Q3.5
explain the results



